### PR TITLE
[AutoTVM] Change to record.pick_best to prevent overwriting previous results.

### PR DESCRIPTION
**Pull request by jwfromm**
I've made a small change to how autotvm.record.pick_best works to prevent it from accidentally deleting autotuning results. Previously, the output file would be overwritten with the best records of the input file. Although this is fine in many cases, if the output file already exists and contains records itself (as might be the case if an autotuning script is run a few times), it is actually much better to choose the best records from both the input and output files. This update simply attaches the input and output records and finds the best records from both to write.

### CODE REVIEW

File: python/tvm/autotvm/record.py [hunk 1]
```diff
@@ -246,7 +250,12 @@ def pick_best(in_file, out_file):
     out_file: str or file
         The filename of output
     """
-    best_context = ApplyHistoryBest(load_from_file(in_file))
+    context = load_from_file(in_file)
+    if os.path.isfile(out_file):
+        out_context = load_from_file(out_file)
+        context = itertools.chain(context, out_context)
```
Comment (huajsj) [inline python/tvm/autotvm/record.py]: for a border condition, when the existing out_file is a old file and coming from a different hardware tune,  if we chain here and choose best later, is that may cause a hidden logic issue?

Comment (jwfromm) [reply python/tvm/autotvm/record.py]: It wont cause any issues but the resulting file will have schedules for both the old hardware and the new hardware. When using that file as a history, only the relevant schedules to the current build will be used. I argue this is desirable behavior as it prevents the old hardware jobs from being totally lost.

File: python/tvm/autotvm/record.py [hunk 2]
```diff
@@ -258,7 +267,7 @@ def pick_best(in_file, out_file):
     logger.info("Extract %d best records from the %s", len(best_set), in_file)
     fout = open(out_file, 'w') if isinstance(out_file, str) else out_file
 
-    for inp, res in load_from_file(in_file):
+    for inp, res in context_clone:
```
Comment (merrymercy) [inline python/tvm/autotvm/record.py]: Why do we need a clone?

Comment (jwfromm) [reply python/tvm/autotvm/record.py]: Generators can only be iterated through once and ApplyBestHistory uses up context. We need a fresh generator during the actual write. We either have to use a tee clone or recreate the generator. I thought this way was more concise.

**Pull request merged**